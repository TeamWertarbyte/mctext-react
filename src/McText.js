import React from 'react'
import PropTypes from 'prop-types'
import defaultColormap from './defaultColormap'
import convertTextToJson from './textToJson'
import TextComponent from './TextComponent'
import { isString } from './util'

export default function McText (props) {
  const {
    prefix,
    children,
    colormap,
    randomChars,
    style,
    ...other
  } = props

  const component = isString(children) ? convertTextToJson(children, prefix) : children

  return (
    <span
      {...other}
      style={{ fontFamily: 'monospace', ...style }}
    >
      <TextComponent
        colormap={colormap}
        component={component}
        randomChars={randomChars}
      />
    </span>
  )
}

McText.propTypes = {
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  colormap: PropTypes.shape({
    black: PropTypes.string,
    dark_blue: PropTypes.string,
    dark_green: PropTypes.string,
    dark_aqua: PropTypes.string,
    dark_red: PropTypes.string,
    dark_purple: PropTypes.string,
    gold: PropTypes.string,
    gray: PropTypes.string,
    dark_gray: PropTypes.string,
    blue: PropTypes.string,
    green: PropTypes.string,
    aqua: PropTypes.string,
    red: PropTypes.string,
    light_purple: PropTypes.string,
    yellow: PropTypes.string,
    white: PropTypes.string
  }),
  randomChars: PropTypes.string,
  prefix: PropTypes.string,
  style: PropTypes.object
}

McText.defaultProps = {
  prefix: 'ยง',
  colormap: defaultColormap,
  randomChars: 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!ยง$%&?#'
}
