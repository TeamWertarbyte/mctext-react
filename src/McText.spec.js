/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'
import McText from './McText'

// mock Math.random to a deterministic random generator
Math.random = jest.fn((() => {
  let x = 0
  return () => {
    x += 0.1
    if (x >= 1) {
      x = 0
    }
    return x
  }
})())

test('Formatting codes text renders properly', () => {
  const tree = renderer.create(
    <McText>§4This is red and this is §ksecret§r and §lbold. Note that §9color codes reset format codes, as you §ocan see.</McText>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Formatting codes text with custom prefix renders properly', () => {
  const tree = renderer.create(
    <McText prefix={'&'}>{'&4This is red and this is &ksecret&r and &lbold. Note that &9color codes reset format codes, as you &ocan see.'}</McText>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('JSON text component renders properly', () => {
  const tree = renderer.create(
    <McText>
      {{'text': '', 'extra': [{'color': 'dark_red', 'obfuscated': true, 'text': 'Wirtschafts-Server mit RPG und klarem Konzept'}, {'color': 'green', 'text': ' | '}, {'color': 'dark_red', 'text': 'SKY-Server '}, {'color': 'green', 'text': '| '}, {'color': 'dark_red', 'text': 'Minigames '}, {'color': 'green', 'text': '| '}, {'color': 'dark_red', 'text': 'Creative '}, {'color': 'green', 'text': '| '}, {'color': 'red', 'text': 'Community von 8 - 66 Jahren '}, {'color': 'green', 'text': '| '}, {'color': 'dark_red', 'text': 'Aktive Admins'}]}}
    </McText>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

test('JSON text component with string extra components renders properly', () => {
  const tree = renderer.create(
    <McText>
      {{ text: 'This ', 'extra': ['works ', { color: 'red', text: 'great!' }] }}
    </McText>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
