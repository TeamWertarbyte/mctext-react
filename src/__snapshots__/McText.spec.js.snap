// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formatting codes text renders properly 1`] = `
<span
  style={
    Object {
      "fontFamily": "monospace",
    }
  }
>
  <span
    style={
      Object {
        "color": undefined,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "textDecoration": undefined,
      }
    }
  >
    
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      This is red and this is 
    </span>
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      <span>
        bipw3!
      </span>
    </span>
    <span
      style={
        Object {
          "color": undefined,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
       and 
    </span>
    <span
      style={
        Object {
          "color": undefined,
          "fontStyle": undefined,
          "fontWeight": "bold",
          "textDecoration": undefined,
        }
      }
    >
      bold. Note that 
    </span>
    <span
      style={
        Object {
          "color": "#5555ff",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      color codes reset format codes, as you 
    </span>
    <span
      style={
        Object {
          "color": "#5555ff",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      can see.
    </span>
  </span>
</span>
`;

exports[`Formatting codes text with custom prefix renders properly 1`] = `
<span
  style={
    Object {
      "fontFamily": "monospace",
    }
  }
>
  <span
    style={
      Object {
        "color": undefined,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "textDecoration": undefined,
      }
    }
  >
    
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      This is red and this is 
    </span>
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      <span>
        #AGNUb
      </span>
    </span>
    <span
      style={
        Object {
          "color": undefined,
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
       and 
    </span>
    <span
      style={
        Object {
          "color": undefined,
          "fontStyle": undefined,
          "fontWeight": "bold",
          "textDecoration": undefined,
        }
      }
    >
      bold. Note that 
    </span>
    <span
      style={
        Object {
          "color": "#5555ff",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      color codes reset format codes, as you 
    </span>
    <span
      style={
        Object {
          "color": "#5555ff",
          "fontStyle": "italic",
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      can see.
    </span>
  </span>
</span>
`;

exports[`JSON text component renders properly 1`] = `
<span
  style={
    Object {
      "fontFamily": "monospace",
    }
  }
>
  <span
    style={
      Object {
        "color": undefined,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "textDecoration": undefined,
      }
    }
  >
    
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      <span>
        ipw3!#AGNUbipw3!#AGNUbipw3!#AGNUbipw3!#AGNUbi
      </span>
    </span>
    <span
      style={
        Object {
          "color": "#55ff55",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
       | 
    </span>
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      SKY-Server 
    </span>
    <span
      style={
        Object {
          "color": "#55ff55",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      | 
    </span>
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      Minigames 
    </span>
    <span
      style={
        Object {
          "color": "#55ff55",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      | 
    </span>
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      Creative 
    </span>
    <span
      style={
        Object {
          "color": "#55ff55",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      | 
    </span>
    <span
      style={
        Object {
          "color": "#ff5555",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      Community von 8 - 66 Jahren 
    </span>
    <span
      style={
        Object {
          "color": "#55ff55",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      | 
    </span>
    <span
      style={
        Object {
          "color": "#aa0000",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      Aktive Admins
    </span>
  </span>
</span>
`;

exports[`JSON text component with string extra components renders properly 1`] = `
<span
  style={
    Object {
      "fontFamily": "monospace",
    }
  }
>
  <span
    style={
      Object {
        "color": undefined,
        "fontStyle": undefined,
        "fontWeight": undefined,
        "textDecoration": undefined,
      }
    }
  >
    This 
    <span>
      works 
    </span>
    <span
      style={
        Object {
          "color": "#ff5555",
          "fontStyle": undefined,
          "fontWeight": undefined,
          "textDecoration": undefined,
        }
      }
    >
      great!
    </span>
  </span>
</span>
`;
